Support Hardware Ledger Auxiliary Seed Generation Method

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#1232 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- 7ea19ff729839bdc9203bd610649f2edb9af7f61
  Add auxiliary seed generation method for Hardware Ledger
  This implement the firmware generation that is happening on Hardware Ledger Nano S and
Nano X. Unfortunately, this process isn't really documented anywhere, and the implementation
has only been tested superficially from a single known mnemonic. I'll try out again with
an actual device as soon as I get my hand on one


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
Add auxiliary seed generation method for Hardware Ledger

This implement the firmware generation that is happening on Hardware Ledger Nano S and
Nano X. Unfortunately, this process isn't really documented anywhere, and the implementation
has only been tested superficially from a single known mnemonic. I'll try out again with
an actual device as soon as I get my hand on one

rvl commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Icarus.hs:
```diff
@@ -151,6 +172,84 @@ generateKeyFromSeed
+    normalizeEd25519 bytes =
+        let
+            (firstByte, rest) = BS.splitAt 1 bytes
+            (rest', lastByte) = BS.splitAt 30 rest
```
We could do this to remove the need for `BS.index`.
```suggestion
            Just (firstByte, rest) = BS.uncons bytes
            Just (rest', lastByte) = BS.unsnoc rest
```
Then put them back together with `BS.cons` and `BS.snoc` to remove the need for `mconcat` and `singleton`.

rvl commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Icarus.hs:
```diff
@@ -151,6 +172,84 @@ generateKeyFromSeed
+    prv <- left show $ xprv $ kL <> kR <> pA <> cc
+    pure $ IcarusKey (xPrvChangePass (mempty :: ByteString) pwd prv)
+  where
+    unsafeFromRight :: (HasCallStack, Show e) => Either e a -> a
```
I think `e` is always `String` here, because `show` has already been applied.
And I don't think that HasCallStack will help if the top-level function doesn't have that constraint.
You could add a comment that any Left values here would be a result of a bug in this function.

KtorZ replied:
True :+1:

rvl commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Icarus.hs:
```diff
@@ -151,6 +172,84 @@ generateKeyFromSeed
+        scalar <- left show $ eitherCryptoError $ Ed25519.scalarDecodeLong bytes
+        pure $ Ed25519.pointEncode $ Ed25519.toPoint scalar
+
+    pbkdf2HmacSha512 :: ByteString -> ByteString
```
Add a comment referencing [BIP39 From mnemonic to seed](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#From_mnemonic_to_seed).

rvl commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Icarus.hs:
```diff
@@ -151,6 +172,84 @@ generateKeyFromSeed
+    -- As described in:
+    --
+    -- "BIP32-Ed25519 Hierarchical Deterministic Keys over a Non-linear Keyspace".
+    normalizeEd25519 :: ByteString -> ByteString
```
It's called buffer "pruning" in [RFC 8032 sec 5.1.5. Key Generation](https://tools.ietf.org/html/rfc8032#section-5.1.5), so it might be better to call it "prune" rather than "normalize".

KtorZ replied:
I kept the same terminology as the Rust guys. But I am happy to follow an RFC :)

rvl commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Icarus.hs:
```diff
@@ -151,6 +172,84 @@ generateKeyFromSeed
 generateKeyFromSeed = unsafeGenerateKeyFromSeed
 
+-- | Hardware Ledger devices generates keys from mnemonic using a different
+-- approach (different from the rest of Cardano).
```
The approach for generating the master key and chain code is called [SLIP-0010](https://github.com/satoshilabs/slips/blob/master/slip-0010.md), right?

KtorZ replied:
Damn. I've been my head through these documents for days, going from one to another and trying to figure out which one applied to the Hardware Ledger that I lost track of them at some point but, I think you're right!

rvl commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Icarus.hs:
```diff
@@ -151,6 +172,84 @@ generateKeyFromSeed
+
+    hashRepeatedly :: ByteString -> (ByteString, ByteString)
+    hashRepeatedly bytes = case BS.splitAt 32 (hmacSha512 bytes) of
+        (kL, kR) | testBit (kL `BS.index` 31) 5 -> hashRepeatedly (kL <> kR)
```
Perhaps we could clarify the guard condition by putting it into a `where` clause and calling it `isInvalidKey`.

From SLIP-0010:

> To avoid invalid master keys, the algorithm is retried with the intermediate hash as new seed if the key is invalid.
> ....
> For ed25519, the last step always succeeds since every 256-bit number (even 0) is a valid private key.

Given that, what is this testBit for? It excludes keys which are multiples of 32 ... are these known to be weak keys?

KtorZ replied:
To be honest. I have no idea and I couldn't find any satisfactory piece of documentation about this. Only thing I know is that the Ledger implementation seems to do this, for... a mysterious reason.

rvl commented on lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivation/IcarusSpec.hs:
```diff
@@ -135,6 +179,30 @@ goldenAddressGeneration test = it title $ do
+
+        forM_ (zip [0..] addrs) $ \(ix, addr) -> do
+            let addrXPrv = deriveAddr (toEnum ix)
+            base58 (paymentAddress @'Mainnet $ publicKey addrXPrv) `shouldBe` addr
```
Good - Base58 for golden tests to match the source data.

rvl reviewed (commented):
I reviewed this by referencing the specs rather than the Python code.

It seems good, except that I'm not sure why `hashRepeatedly` is required. Does the Ledger also do this?

It would help with understanding if the `generateKeyFromHardwareLedger` function had comments and naming that referenced the BIPs, SLIPs, and RFCs that are used.

KtorZ committed:
revise generateKeyFromHardwareLedger implementation

- Add comments to reference known specifications
- Change names to match those use in RFC and papers
- Favor cons/snoc from the ByteString library for readability

rvl reviewed (approved):
Looks good. If it matches the Ledger implementation then good. It would be nice to eventually know the explanation for `isInvalidKey`.

KtorZ committed:
add additional golden tests for Ledger and Trezor wallets
